### PR TITLE
[Program: GCI] (Comment task - Part 2/2) Prepare comment task endpoints [Mentorship - Backend]

### PR DESCRIPTION
**Pull request by asawesome07**
### Description
Added Task comment model and modified task_dao for task_comment

### Type of Change:
- Code

**Code/Quality Assurance Only**
- New feature (non-breaking change which adds functionality pre-approved by mentors)

### How Has This Been Tested?
I tested this using `python -m unittest discover tests`

Here's a screenshot
Capture

### Checklist:
- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-review of my own code or materials
- [x] I have commented my code or provided relevant documentation, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] Any dependent changes have been merged
- [x] Update Postman API at /docs folder
- [x] Update Swagger documentation and the exported file at /docs folder

**Code/Quality Assurance Only**
- [x] My changes generate no new warnings 
- [ ] My PR currently breaks something (fix or feature that would cause existing functionality to not work as expected)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been published in downstream modules

### CODE REVIEW

File: app/database/models/task_comment.py [hunk 1]
```diff
@@ -83,3 +147,28 @@ def delete_from_db(self):
...
+
+    ID = 'id'
+    COMMENT = 'comment'
+    USER_ID = 'user_id'
+    TASK_ID = 'task_id'
+    CREATION_DATE = 'creation_date'
+    MODIFICATION_DATE = 'modification_date'
+
+    def values(self):
+        """Returns a list containing a task."""
```
Comment (annabauza) [inline app/database/models/task_comment.py]: List of what?

File: app/database/models/task_comment.py [hunk 2]
```diff
@@ -83,3 +147,28 @@ def delete_from_db(self):
         """Deletes the record of task comment from the database."""
         db.session.delete(self)
         db.session.commit()
+
+
+@unique
+class TaskCommentsFields(Enum):
+    """Represents a task comment attributes' name.
+
+    Attributes:
```
Comment (annabauza) [inline app/database/models/task_comment.py:148]: Not sure if those are attributes. Do we use this convention in other enums?

File: app/database/models/task_comment.py [hunk 3]
```diff
@@ -19,60 +21,122 @@ class TaskCommentModel(db.Model):
...
+        self.task_comments = new_list
+        self.save_to_db()
+
+    def is_empty(self):
         """Returns True if the task comment model is empty, and False otherwise."""
-        return cls.query.first() is None
+        return len(self.task_comments) == 0
+
+    @classmethod
+    def find_by_id(cls, _id):
```
Comment (annabauza) [inline app/database/models/task_comment.py]: shell we add mentorship_relation_id here or this is unique?

File: app/api/dao/task_comment.py [hunk 4]
```diff
@@ -28,47 +32,35 @@ def create_task_comment(user_id, task_id, data):
             response code.
```
Comment (annabauza) [inline app/api/dao/task_comment.py:29]: can we add error response to docs?

File: app/api/dao/task_comment.py [hunk 5]
```diff
@@ -1,21 +1,25 @@
...
+    def create_task_comment(relation_id, user_id, task_id, data):
         """Creates a new task.
 
         Creates a new task comment,
 
         Args:
+            relation_id: The id of the mentorship relation.
             user_id: The id of the user.
             task_id: The id of the task relation.
             data: A list containing the comment of the task comment.
```
Comment (annabauza) [inline app/api/dao/task_comment.py]: This sentence is unclear.

File: app/api/dao/task_comment.py [hunk 6]
```diff
@@ -28,47 +32,35 @@ def create_task_comment(user_id, task_id, data):
...
-            if retrieving the task comment was successful or not as a string.
-            The last element is the HTTP
-            response code.
-        """
+        task = relation.tasks_list.find_task_by_id(task_id)
+        if task is None:
+            return messages.TASK_DOES_NOT_EXIST, 404
 
-        task_comment = TaskCommentModel.find_by_id(_id)
+        comment = data['comment']
```
Comment (annabauza) [inline app/api/dao/task_comment.py]: is it single comment or multiple comments? I think name is confusing

File: app/api/dao/task_comment.py [hunk 7]
```diff
@@ -79,21 +71,35 @@ def get_task_comments_by_task_id(task_id):
...
 
-        task_comment = TaskCommentModel.find_by_task_id(task_id)
+        relation = MentorshipRelationModel.find_by_id(_id=relation_id)
+        if relation is None:
+            return messages.MENTORSHIP_RELATION_DOES_NOT_EXIST, 404
 
-        if task_comment:
-            return task_comment, 200
+        if relation.state != MentorshipRelationState.ACCEPTED:
+            return messages.UNACCEPTED_STATE_RELATION, 400
```
Comment (annabauza) [inline app/api/dao/task_comment.py:73]: this return is not documented

File: app/api/dao/task_comment.py [hunk 8]
```diff
@@ -79,21 +71,35 @@ def get_task_comments_by_task_id(task_id):
...
 
-        return messages.TASK_COMMENT_DOES_NOT_EXIST, 404
+        task = relation.tasks_list.find_task_by_id(task_id)
+        if task is None:
+            return messages.TASK_DOES_NOT_EXIST, 404
```
Comment (annabauza) [inline app/api/dao/task_comment.py:77]: this return is not documented

File: app/api/dao/task_comment.py [hunk 9]
```diff
@@ -79,21 +71,35 @@ def get_task_comments_by_task_id(task_id):
...
+
+        if not (user_id == relation.mentee_id or user_id == relation.mentor_id):
+            return messages.USER_NOT_INVOLVED_IN_THIS_MENTOR_RELATION, 401
```
Comment (annabauza) [inline app/api/dao/task_comment.py:80]: this return is not documented

File: app/api/dao/task_comment.py [hunk 10]
```diff
@@ -106,12 +112,27 @@ def modify_comment(user_id, _id, comment):
             response code.
         """
 
-        task_comment = TaskCommentModel.find_by_id(_id)
+        relation = MentorshipRelationModel.find_by_id(_id=relation_id)
+        if relation is None:
+            return messages.MENTORSHIP_RELATION_DOES_NOT_EXIST, 404
```
Comment (annabauza) [inline app/api/dao/task_comment.py:111]: this return is not documented

File: app/api/dao/task_comment.py [hunk 11]
```diff
@@ -106,12 +112,27 @@ def modify_comment(user_id, _id, comment):
...
+
+        if relation.state != MentorshipRelationState.ACCEPTED:
+            return messages.UNACCEPTED_STATE_RELATION, 400
```
Comment (annabauza) [inline app/api/dao/task_comment.py:114]: this return is not documented

File: app/api/dao/task_comment.py [hunk 12]
```diff
@@ -106,12 +112,27 @@ def modify_comment(user_id, _id, comment):
...
+
+        task = relation.tasks_list.find_task_by_id(task_id)
+        if task is None:
+            return messages.TASK_DOES_NOT_EXIST, 404
```
Comment (annabauza) [inline app/api/dao/task_comment.py:118]: this return is not documented

File: app/api/dao/task_comment.py [hunk 13]
```diff
@@ -106,12 +112,27 @@ def modify_comment(user_id, _id, comment):
...
+
+        if not (user_id == relation.mentee_id or user_id == relation.mentor_id):
+            return messages.USER_NOT_INVOLVED_IN_THIS_MENTOR_RELATION, 401
```
Comment (annabauza) [inline app/api/dao/task_comment.py:121]: this return is not documented

File: app/api/dao/task_comment.py [hunk 14]
```diff
@@ -137,11 +160,27 @@ def delete_comment(user_id, _id):
             response code.
         """
 
-        task_comment = TaskCommentModel.find_by_id(_id)
+        relation = MentorshipRelationModel.find_by_id(_id=relation_id)
+        if relation is None:
+            return messages.MENTORSHIP_RELATION_DOES_NOT_EXIST, 404
```
Comment (annabauza) [inline app/api/dao/task_comment.py:159]: this return is not documented

File: app/api/dao/task_comment.py [hunk 15]
```diff
@@ -137,11 +160,27 @@ def delete_comment(user_id, _id):
...
+
+        if relation.state != MentorshipRelationState.ACCEPTED:
+            return messages.UNACCEPTED_STATE_RELATION, 400
```
Comment (annabauza) [inline app/api/dao/task_comment.py:162]: this return is not documented

File: app/api/dao/task_comment.py [hunk 16]
```diff
@@ -137,11 +160,27 @@ def delete_comment(user_id, _id):
...
+
+        task = relation.tasks_list.find_task_by_id(task_id)
+        if task is None:
+            return messages.TASK_DOES_NOT_EXIST, 404
```
Comment (annabauza) [inline app/api/dao/task_comment.py:166]: this return is not documented

File: app/api/dao/task_comment.py [hunk 17]
```diff
@@ -137,11 +160,27 @@ def delete_comment(user_id, _id):
...
+
+        if not (user_id == relation.mentee_id or user_id == relation.mentor_id):
+            return messages.USER_NOT_INVOLVED_IN_THIS_MENTOR_RELATION, 401
```
Comment (annabauza) [inline app/api/dao/task_comment.py:169]: this return is not documented

File: app/api/resources/mentorship_relation.py [hunk 18]
```diff
@@ -335,3 +336,123 @@ def put(cls, request_id, task_id):
...
+    @jwt_required
+    @mentorship_relation_ns.doc('create_task_comment_in_task_in_mentorship_relation')
+    @mentorship_relation_ns.expect(auth_header_parser, create_task_comment_body)
+    @mentorship_relation_ns.response(200, 'Task comment was created successfully.')
+    def post(cls, request_id, task_id):
+        """
+        Create a task comment.
+        """
+
+        # TODO check if user id is well parsed, if it is an integer
```
Comment (annabauza) [inline app/api/resources/mentorship_relation.py:357]: you intended to add something here?

File: app/api/resources/mentorship_relation.py [hunk 19]
```diff
@@ -335,3 +336,123 @@ def put(cls, request_id, task_id):
...
+
+    @classmethod
+    @jwt_required
+    @mentorship_relation_ns.doc('list_tasks_in_mentorship_relation')
+    @mentorship_relation_ns.expect(auth_header_parser)
+    @mentorship_relation_ns.response(200, 'List task comments from a task from a mentorship relation with success.',
+                                     model=list_task_comment_response_body)
+    def get(cls, request_id, task_id):
+        """
+        List all task comment from a task from a mentorship relation.
```
Comment (annabauza) [inline app/api/resources/mentorship_relation.py]: comments not comment

File: app/api/resources/mentorship_relation.py [hunk 20]
```diff
@@ -335,3 +336,123 @@ def put(cls, request_id, task_id):
...
+        """
+
+        # TODO check if user id is well parsed, if it is an integer
```
Comment (annabauza) [inline app/api/resources/mentorship_relation.py:397]: again is something missing here?

File: app/api/resources/mentorship_relation.py [hunk 21]
```diff
@@ -335,3 +336,123 @@ def put(cls, request_id, task_id):
...
+
+
+@mentorship_relation_ns.route('mentorship_relation/<int:request_id>/task/<int:task_id>/taskcomment/<int:taskcomment_id>')
+class DeleteTaskComment(Resource):
+
+    @classmethod
+    @jwt_required
+    @mentorship_relation_ns.doc('delete_task_in_mentorship_relation')
+    @mentorship_relation_ns.expect(auth_header_parser)
+    @mentorship_relation_ns.response(200, 'Task comment was deleted successfully')
```
Comment (annabauza) [inline app/api/resources/mentorship_relation.py:415]: delete_comment returns some new codes, should be documented

File: app/api/resources/mentorship_relation.py [hunk 22]
```diff
@@ -335,3 +336,123 @@ def put(cls, request_id, task_id):
...
+            return marshal(response, list_task_comment_response_body), 200
+
+
+@mentorship_relation_ns.route('mentorship_relation/<int:request_id>/task/<int:task_id>/taskcomment/<int:taskcomment_id>')
+class DeleteTaskComment(Resource):
+
+    @classmethod
+    @jwt_required
+    @mentorship_relation_ns.doc('delete_task_in_mentorship_relation')
+    @mentorship_relation_ns.expect(auth_header_parser)
```
Comment (annabauza) [inline app/api/resources/mentorship_relation.py:414]: Shouldn't we add some arguments to documentation?

File: app/api/resources/mentorship_relation.py [hunk 23]
```diff
@@ -335,3 +336,123 @@ def put(cls, request_id, task_id):
...
+        return {}
+
+
+@mentorship_relation_ns.route('mentorship_relation/<int:request_id>/task/<int:task_id>/taskcommnents')
+class ListTasksComment(Resource):
+
+    @classmethod
+    @jwt_required
+    @mentorship_relation_ns.doc('list_tasks_in_mentorship_relation')
+    @mentorship_relation_ns.expect(auth_header_parser)
```
Comment (annabauza) [inline app/api/resources/mentorship_relation.py:385]: Shouldn't we add some arguments to documentation?

File: app/api/resources/mentorship_relation.py [hunk 24]
```diff
@@ -335,3 +336,123 @@ def put(cls, request_id, task_id):
...
+    @mentorship_relation_ns.response(200, 'List task comments from a task from a mentorship relation with success.',
+                                     model=list_task_comment_response_body)
```
Comment (annabauza) [inline app/api/resources/mentorship_relation.py:387]: it returns some more codes, please fill up.

File: app/api/resources/mentorship_relation.py [hunk 25]
```diff
@@ -335,3 +336,123 @@ def put(cls, request_id, task_id):
...
+        return response
+
+
+@mentorship_relation_ns.route('mentorship_relation/<int:request_id>/task/<int:task_id>/taskcommnent/<int:taskcomment_id>/modify')
+class ModifyTaskComment(Resource):
+
+    @classmethod
+    @jwt_required
+    @mentorship_relation_ns.doc('update_task_comment_in_task_in_mentorship_relation')
+    @mentorship_relation_ns.expect(auth_header_parser, create_task_comment_body)
```
Comment (annabauza) [inline app/api/resources/mentorship_relation.py:437]: Shouldn't we add some arguments to documentation?

File: app/api/resources/mentorship_relation.py [hunk 26]
```diff
@@ -335,3 +336,123 @@ def put(cls, request_id, task_id):
...
         return response
+
+
+@mentorship_relation_ns.route('mentorship_relation/<int:request_id>/task/<int:task_id>/taskcommnent')
+class CreateTaskComment(Resource):
+
+    @classmethod
+    @jwt_required
+    @mentorship_relation_ns.doc('create_task_comment_in_task_in_mentorship_relation')
+    @mentorship_relation_ns.expect(auth_header_parser, create_task_comment_body)
```
Comment (annabauza) [inline app/api/resources/mentorship_relation.py:347]: Shouldn't we add some arguments to documentation?

File: app/api/resources/mentorship_relation.py [hunk 27]
```diff
@@ -335,3 +336,123 @@ def put(cls, request_id, task_id):
...
+    @jwt_required
+    @mentorship_relation_ns.doc('update_task_comment_in_task_in_mentorship_relation')
+    @mentorship_relation_ns.expect(auth_header_parser, create_task_comment_body)
+    @mentorship_relation_ns.response(200, 'Task comment was modified successfully')
+    def put(cls, request_id, task_id, taskcomment_id):
+        """
+        Modify a task comment.
+        """
+
+        # TODO check if user id is well parsed, if it is an integer
```
Comment (annabauza) [inline app/api/resources/mentorship_relation.py:446]: is something missing here?

File: app/api/resources/mentorship_relation.py [hunk 28]
```diff
@@ -335,3 +336,123 @@ def put(cls, request_id, task_id):
...
+
+
+@mentorship_relation_ns.route('mentorship_relation/<int:request_id>/task/<int:task_id>/taskcommnent/<int:taskcomment_id>/modify')
+class ModifyTaskComment(Resource):
+
+    @classmethod
+    @jwt_required
+    @mentorship_relation_ns.doc('update_task_comment_in_task_in_mentorship_relation')
+    @mentorship_relation_ns.expect(auth_header_parser, create_task_comment_body)
+    @mentorship_relation_ns.response(200, 'Task comment was modified successfully')
```
Comment (annabauza) [inline app/api/resources/mentorship_relation.py:438]: this returns some more codes, not only 200.

**Review by annabauza — CHANGES_REQUESTED** [review]
hey left you some comments, mostly minor comments and docs :)

File: app/database/models/task_comment.py (hunk 1 above)
Comment (asawesome07) [reply app/database/models/task_comment.py]: fixed

File: app/database/models/task_comment.py (hunk 2 above)
Comment (asawesome07) [reply app/database/models/task_comment.py:148]: yes, this convention is also used to describe a task in tasks_list model.
app/database/models/tasks_list.py

File: app/database/models/task_comment.py (hunk 3 above)
Comment (asawesome07) [reply app/database/models/task_comment.py]: removed the function as it wasn't used

File: app/api/resources/mentorship_relation.py (hunk 19 above)
Comment (asawesome07) [reply app/api/resources/mentorship_relation.py]: fixed

File: app/api/dao/task_comment.py (hunk 5 above)
Comment (asawesome07) [reply app/api/dao/task_comment.py]: fixed

File: app/api/dao/task_comment.py (hunk 6 above)
Comment (asawesome07) [reply app/api/dao/task_comment.py]: fixed

File: app/api/dao/task_comment.py (hunk 5 above)
Comment (annabauza) [reply app/api/dao/task_comment.py]: comment of the task comment?

File: app/api/dao/task_comment.py (hunk 6 above)
Comment (annabauza) [reply app/api/dao/task_comment.py]: have you pushed changes?

**Review by annabauza — APPROVED** [review]
well done :)

**Comment by annabauza** [comment]
Can you please merge this PR, so we have everything in one PR? I don't have rights to this repo. Thank you in advance.

**Comment by asawesome07** [comment]
> Can you please merge this PR, so we have everything in one PR? I don't have rights to this repo. Thank you in advance.

done

**Pull request merged**